internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/cmd"
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const (
+	requestTimeout    = 10 * time.Second
+	maxRequestEntries = 100
+)
+
+type Client struct {
+	options *ClientOptions
+
+	targetName string
+	remoteURL  string
+	entries    []lokiEntryWithService
+
+	httpClient *http.Client
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return NewClientWithOptions(target, &ClientOptions{})
+}
+
+// ClientOptions allows overriding default parameters (e.g. for testing)
+type ClientOptions struct {
+	RequestTimeout    time.Duration
+	MaxRequestEntries int
+}
+
+func NewClientWithOptions(target *plan.LogTarget, options *ClientOptions) *Client {
+	options = fillDefaultOptions(options)
+	return &Client{
+		options:    options,
+		targetName: target.Name,
+		remoteURL:  target.Location,
+		httpClient: &http.Client{Timeout: options.RequestTimeout},
+	}
+}
+
+func fillDefaultOptions(options *ClientOptions) *ClientOptions {
+	if options.RequestTimeout == 0 {
+		options.RequestTimeout = requestTimeout
+	}
+	if options.MaxRequestEntries == 0 {
+		options.MaxRequestEntries = maxRequestEntries
+	}
+	return options
+}
+
+func (c *Client) Add(entry servicelog.Entry) error {
+	if n := len(c.entries); n >= c.options.MaxRequestEntries {
+		// make room for 1 entry
+		c.entries = c.entries[(n - c.options.MaxRequestEntries + 1):]
```

Comment:
The element that is being pushed out needs to be zeroed so data may be garbage collected.

Reply:
can do. I also realised that it's always element 0 that we're pushing out, so I've simplified the indexing logic here.